File: backend/GraficaModerna.Domain/Entities/Product.cs
using System.ComponentModel.DataAnnotations;

namespace GraficaModerna.Domain.Entities;

public class Product(
    string name,
    string description,
    decimal price,
    List<string> imageUrls,
    decimal weight,
    int width,
    int height,
    int length,
    int stockQuantity) : BaseEntity
{
    public string Name { get; private set; } = name;
    public string Description { get; private set; } = description;
    public decimal Price { get; private set; } = price;
    public List<string> ImageUrls { get; private set; } = imageUrls;
    public decimal Weight { get; private set; } = weight;
    public int Width { get; private set; } = width;
    public int Height { get; private set; } = height;
    public int Length { get; private set; } = length;
    public int StockQuantity { get; private set; } = stockQuantity;
    public bool IsActive { get; private set; } = true;

    [Timestamp] public byte[]? RowVersion { get; set; }

    public void Update(string name, string description, decimal price, List<string> imageUrls, decimal weight, int width,
        int height, int length, int stockQuantity)
    {
        Name = name;
        Description = description;
        Price = price;
        ImageUrls = imageUrls;
        Weight = weight;
        Width = width;
        Height = height;
        Length = length;
        StockQuantity = stockQuantity;
    }

    public void DebitStock(int quantity)
    {
        if (quantity < 0) throw new ArgumentException("Quantidade inválida.");

        if (StockQuantity < quantity)
            throw new InvalidOperationException($"Estoque insuficiente para o produto '{Name}'.");

        StockQuantity -= quantity;
    }

    public void ReplenishStock(int quantity)
    {
        StockQuantity += quantity;
    }

    public void Deactivate()
    {
        IsActive = false;
    }
}

File: app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Concerns\HasUuids;

class Product extends Model
{
    use HasUuids;

    protected $fillable = [
        'name', 'description', 'price', 'image_urls', 
        'weight', 'width', 'height', 'length', 'stock_quantity', 'is_active'
    ];

    protected $casts = [
        'image_urls' => 'array',
        'price' => 'decimal:2',
        'weight' => 'decimal:3',
        'is_active' => 'boolean',
    ];

    public function debitStock(int $quantity)
    {
        if ($quantity < 0) throw new \InvalidArgumentException("Quantidade inválida.");
        
        if ($this->stock_quantity < $quantity) {
            throw new \Exception("Estoque insuficiente para o produto '{$this->name}'.");
        }

        $this->decrement('stock_quantity', $quantity);
    }

    public function replenishStock(int $quantity)
    {
        $this->increment('stock_quantity', $quantity);
    }

    public function deactivate()
    {
        $this->update(['is_active' => false]);
    }
}